yarn-project/p2p/src/services/peer-manager/peer_manager.ts

```diff
@@ -670,6 +727,71 @@ export class PeerManager implements PeerManagerInterface {
     }
   }
 
+  /**
+   * Performs auth Handshake with the Peer
+   * A superset of the status handshake. Also includes a challenge that needs to be signed by the peer's validator key.
+   * @param: peerId The Id of the peer to request the Status from.
+   * */
+  private async exchangeAuthHandshake(peerId: PeerId) {
+    try {
+      const ourStatus = await this.createStatusMessage();
+      const authRequest = new AuthRequest(ourStatus, Fr.random());
+
+      //Note: Technically we don't have to send our status to peer as well, but we do.
+      //It will be easier to update protocol in the future this way if need be.
+      this.logger.debug(`Initiating auth handshake with peer ${peerId}`);
+      const { status, data } = await this.reqresp.sendRequestToPeer(
+        peerId,
+        ReqRespSubProtocol.AUTH,
+        authRequest.toBuffer(),
+      );
+      const logData = { peerId, status: ReqRespStatus[status], data: data ? bufferToHex(data) : undefined };
+      if (status !== ReqRespStatus.SUCCESS) {
+        this.logger.debug(`Disconnecting peer ${peerId} who failed to respond auth handshake`, logData);
+        await this.disconnectPeer(peerId);
+        return;
+      }
+
+      const peerAuthResponse = AuthResponse.fromBuffer(data);
+
+      const peerStatusMessage = peerAuthResponse.status;
+      if (!ourStatus.validate(peerStatusMessage)) {
+        this.logger.warn(`Disconnecting peer ${peerId} due to failed status handshake as part of auth.`, logData);
+        await this.disconnectPeer(peerId);
+        return;
+      }
+
+      const hashToRecover = authRequest.getPayloadToSign();
+      const sender = recoverAddress(hashToRecover, peerAuthResponse.signature);
+      const registeredValidators = await this.epochCache.getRegisteredValidators();
+      const found = registeredValidators.find(v => v.toString() === sender.toString()) !== undefined;
+      if (!found) {
+        this.logger.debug(
+          `Disconnecting peer ${peerId} due to failed auth handshake, peer is not a registered validator.`,
+          {
+            peerId,
+            address: sender.toString(),
+          },
+        );
+        await this.disconnectPeer(peerId);
+        return;
+      }
+
+      const peerIdString = peerId.toString();
+      this.authenticatedPeers.set(peerIdString, sender);
```

Comment:
Should we limit the number of p2p peers that can auth using the same validator key? It seems like a single validator can spawn an arbitrary number of nodes and auth to the supernode using the same key for all of them.

Reply:
Right yes. That is a good idea. I hadn't considered that.